### PR TITLE
Leaderboard Styles Enhancement & Bug Fixes

### PR DESCRIPTION
**Pull request by Cena-Studio**
# Overview

This PR aims to improve the leaderboard style so that it can have a responsive size with the window size changing. Besides, it adds national flag icons to the location column of the entries of the leaderboard. It also fixed a backend bug in `device_controller#raw_data` that causes an error when fetching the data without having any entries in the database.

# Testing

- Run `rails s` to start the local server.
- Access the leaderboard page `http://localhost:3000/leaderboard/grouped` through a browser.
- Resize the browser window to see the responsive resizing of the leaderboard.

### CODE REVIEW

File: app/controllers/device_controller.rb
Comment (megan-godfrey) [inline app/controllers/device_controller.rb:32]: Not important for now but would be good to update this field name at some point

File: app/assets/stylesheets/leaderboard.css
Comment (megan-godfrey) [inline app/assets/stylesheets/leaderboard.css:257]: Not sure where to leave this comment - 
There's currently no padding under the 'g's in the carbon emissions columns when hovering over each entry. If we're keeping the units here (i.e. not moving them to the header) could we increase the top and bottom padding maybe?

image

**Review by megan-godfrey — APPROVED** [review]
The responsive design works nicely, I've just added one very minor comment about padding that may stop being relevant.

File: app/assets/stylesheets/leaderboard.css
Comment (Cena-Studio) [reply app/assets/stylesheets/leaderboard.css:257]: Can I round this by using a smaller font size for the units? Cuz changing the height of the wrapper will affect the layout of other elements in the wrapper

Comment (megan-godfrey) [reply app/assets/stylesheets/leaderboard.css:257]: Sure that sounds good :)

File: app/assets/stylesheets/leaderboard.css
Comment (megan-godfrey) [inline app/assets/stylesheets/leaderboard.css:110]: I'm wondering if we could just show the flags in the location column, without the country code, then make them a little bigger?

Then center all of the contents of the data columns, just leaving the group name left-aligned

My thoughts are 
- this would make the flag easier to see
- centering the data would allow for centering the platform icons while keeping the layout consistent between columns

File: app/controllers/group_controller.rb
Comment (megan-godfrey) [inline app/controllers/group_controller.rb:47]: I assume this was leftover from testing?

**Review by megan-godfrey — CHANGES_REQUESTED** [review]

File: app/assets/stylesheets/leaderboard.css
Comment (Cena-Studio) [reply app/assets/stylesheets/leaderboard.css:110]: Ahh I totally forgot that I was writing the above comment.

File: app/javascript/group/index.js
Comment (megan-godfrey) [inline app/javascript/group/index.js:25]: Left over from testing? I'm getting moosebirds all the way down

**Comment by Womblue** [comment]
Something I'm noticing with this branch is that the leaderboard itself loads a little slower (at least on my laptop, it's noticeable): [Recording](https://github.com/openflighthpc/carbon-leaderboard/assets/57297493/288b06c4-f5b8-4dbb-ad9a-e6411c74ddac), first page is this branch, second page is the public leaderboard server. I feel like this amount of delay isn't a significant issue, but I imagine it'll get worse when more devices are on the leaderboard.

**Comment by megan-godfrey** [comment]
Could be caused by loading the icons? Could try reducing the file sizes of the platform icons. I'm not sure how urgent it is though

**Comment by Womblue** [comment]
I'm going to try and add a bunch of new groups to my local database and see if the issue becomes more serious

File: app/javascript/group/index.js
Comment (Cena-Studio) [reply app/javascript/group/index.js:25]: I'll change my testing approach. It caused too many issues.

File: app/assets/stylesheets/leaderboard.css
Comment (megan-godfrey) [inline app/assets/stylesheets/leaderboard.css:110]: Minor comment - it might make sense to have the platform and flag icons for ranks 1-3 bigger than for the other ranks, as the text is bigger for the top ranks

Comment (Cena-Studio) [reply app/assets/stylesheets/leaderboard.css]: Good idea. I've updated the style.

Comment (megan-godfrey) [reply app/assets/stylesheets/leaderboard.css]: Yeeee that looks nice :)

**Review by megan-godfrey — COMMENTED** [review]
LGTM! Pending whether Matt deems that the loading time has gotten too long...

**Comment by Womblue** [comment]
I'm afraid I'm going to have to say that the loading time IS too long. I added 26 total groups to my own database, and this is the result:

[Screencast from 2024-05-09 17-31-19.webm](https://github.com/openflighthpc/carbon-leaderboard/assets/57297493/97a71f18-33e6-4450-bad0-d7a5d8c9a21b)

I can see for sure that it's caused by the platform icons and NOT the flags, because the individual leaderboard still loads fast (which has the flags, but not the platform icons). In the video, I'm swapping between the individual group page and the individual page to demonstrate the difference.

There's a lot of slowdown there with just 26 groups - the current public leaderboard has 64 groups, and that's the MINIMUM we'll be dealing with, given that users will be adding their own devices to the leaderboard. There must be a way to deal with this, given that the flag images are of a similar size and don't have the same issues.

**Comment by Womblue** [comment]
Found the culprit (I think)
``` 
def platform_icon
    platform = self.platform.downcase
    if Boavizta.type_exists?(self.instance_type, 'alces')
      'alces'
    elsif %w(aws azure openstack).include?(platform)
      platform
    else
      'server'
    end
  end
```
This function is run individually for each item on the leaderboard. In order to check if the device is an alces device, we're pinging Boavizta, which introduces a small delay while we wait for the response.

**Comment by Womblue** [comment]
Just added a couple of commits which fix the slowdown.

* Alces Cloud devices are now checked by Boavizta only when added to the database, instead of being checked every single time a platform string is used.
* With this change, a device's `platform` now has an extra potential option: `"Alces Cloud"`. I've included a data migration to update any existing alces instances - so if you run `rails db:migrate:with_data`, then any instances already in your local database which would have been labelled `Alces Cloud` will have their platforms updated.

File: db/data/20240509174108_set_alces_platform_names.rb
Comment (megan-godfrey) [inline db/data/20240509174108_set_alces_platform_names.rb:5]: Just `find_each` should work - shouldn't also need `each`

**Comment by Womblue** [comment]
LGTM!

**Pull request merged**
